### ISSUE TITLE
AOT Dispatcher converts a single `detach` call to multiple `aten.detach`

### ISSUE THREAD

**Issue by StrongerXi** [pytorch/pytorch#151142]
### 🐛 Describe the bug

I observed this in #150706, where a `detach` in Dynamo graph becomes multiple `aten.alias` in AOT graph, [tlparse](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpqAJWZf/-_0_0_0/compilation_metrics_13.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000):

![Image](https://github.com/user-attachments/assets/04e5e984-00fb-4880-871c-b80fbcac8b0f)

This isn't blocking #150706 because I have a separate fix to eliminate the `detach` in Dynamo graph, but I'm not sure if the AOTDispatcher behavior is intentional, and if it is, feel free to close this issue.

Minimal Repro:
```python
import torch

@torch.compile(backend="aot_eager", fullgraph=True)
def f(x):
    x = x.detach()
    res = x + 1
    return res

f(torch.ones(1))
```

### Error logs

Running with `TORCH_LOGS="graph_code, aot_graph" gives:

```
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code] TRACED GRAPH
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]  ===== pre insert_deferred_runtime_asserts __compiled_fn_1 =====
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]  <eval_with_key>.0 class GraphModule(torch.nn.Module):
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]     def forward(self, L_x_: "f32[1]"):
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]         l_x_ = L_x_
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]          # File: /home/ryanguo99/scratch/compile-time.py:5 in f, code: x = x.detach()
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]         x: "f32[1]" = l_x_.detach();  l_x_ = None
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]          # File: /home/ryanguo99/scratch/compile-time.py:6 in f, code: res = x + 1
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]         res: "f32[1]" = x + 1;  x = None
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]         return (res,)
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]
V0411 16:03:36.439000 2423193 torch/fx/passes/runtime_assert.py:118] [0/0] [__graph_code]
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code] TRACED GRAPH
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]  ===== __compiled_fn_1 =====
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]  /home/ryanguo99/repos/pytorch/torch/fx/_lazy_graph_module.py class GraphModule(torch.nn.Module):
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]     def forward(self, L_x_: "f32[1][1]cpu"):
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]         l_x_ = L_x_
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]          # File: /home/ryanguo99/scratch/compile-time.py:5 in f, code: x = x.detach()
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]         x: "f32[1][1]cpu" = l_x_.detach();  l_x_ = None
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]          # File: /home/ryanguo99/scratch/compile-time.py:6 in f, code: res = x + 1
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]         res: "f32[1][1]cpu" = x + 1;  x = None
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]         return (res,)
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]
V0411 16:03:36.440000 2423193 torch/_dynamo/output_graph.py:1431] [0/0] [__graph_code]
V0411 16:03:36.531000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:123] [0/0] [__aot_graphs] aot_config id: 0, fw_metadata=ViewAndMutationMeta(input_info=[InputAliasInfo(is_leaf=True, mutates_data=False, mutates_metadata=False, mutations_hidden_from_autograd=True, mutations_under_no_grad_or_inference_mode=False, mutation_inductor_storage_resize=False, mutates_storage_metadata=False, requires_grad=False, keep_input_mutations=True)], output_info=[OutputAliasInfo(output_type=<OutputType.non_alias: 1>, raw_type=<class 'torch._subclasses.functional_tensor.FunctionalTensor'>, base_idx=None, dynamic_dims=set(), requires_grad=False, functional_tensor=None)], num_intermediate_bases=0, keep_input_mutations=True, traced_tangents=[], subclass_inp_meta=[PlainTensorMeta(unwrapped_idx=0, memory_format=None)], subclass_fw_graph_out_meta=[PlainTensorMeta(unwrapped_idx=0, memory_format=None)], subclass_tangent_meta=[], is_train=False, traced_tangent_metas=None, num_symints_saved_for_bw=None, grad_enabled_mutation=None, deterministic=None, static_input_indices=[], tokens={}, indices_of_inputs_that_requires_grad_with_mutations_in_bw=[], bw_donated_idxs=None, num_backward_tokens=0, num_graphsafe_rng_states=0, graphsafe_rng_state_index=None),subclass_metadata=None
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs] TRACED GRAPH
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]  ===== Forward graph 0 =====
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]  /home/ryanguo99/repos/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]     def forward(self, arg0_1: "f32[1][1]cpu"):
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]          # File: /home/ryanguo99/scratch/compile-time.py:5 in f, code: x = x.detach()
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]         detach: "f32[1][1]cpu" = torch.ops.aten.detach.default(arg0_1);  arg0_1 = None
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]         detach_1: "f32[1][1]cpu" = torch.ops.aten.detach.default(detach);  detach = None
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]         detach_2: "f32[1][1]cpu" = torch.ops.aten.detach.default(detach_1);  detach_1 = None
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]          # File: /home/ryanguo99/scratch/compile-time.py:6 in f, code: res = x + 1
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]         add: "f32[1][1]cpu" = torch.ops.aten.add.Tensor(detach_2, 1);  detach_2 = None
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]         return (add,)
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]
I0411 16:03:36.537000 2423193 torch/_functorch/_aot_autograd/dispatch_and_compile_graph.py:202] [0/0] [__aot_graphs]
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code] TRACED GRAPH
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]  ===== tensorify_python_scalars =====
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]  /home/ryanguo99/repos/pytorch/torch/fx/_lazy_graph_module.py class <lambda>(torch.nn.Module):
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]     def forward(self, arg0_1: "f32[1]"):
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]          # File: /home/ryanguo99/scratch/compile-time.py:5 in f, code: x = x.detach()
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]         detach: "f32[1]" = torch.ops.aten.detach.default(arg0_1);  arg0_1 = None
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]         detach_1: "f32[1]" = torch.ops.aten.detach.default(detach);  detach = None
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]         detach_2: "f32[1]" = torch.ops.aten.detach.default(detach_1);  detach_1 = None
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]          # File: /home/ryanguo99/scratch/compile-time.py:6 in f, code: res = x + 1
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]         add: "f32[1]" = torch.ops.aten.add.Tensor(detach_2, 1);  detach_2 = None
V0411 16:03:36.537000 2423193 torch/fx/passes/_tensorify_python_scalars.py:364] [0/0] [__graph_code]         return (add,)
```

### Versions

main 1a1a32ce5af, Python 3.11

**Comment by StrongerXi** [comment]
Even if we use `aten.detach` upfront, it repros:

```python
import torch

@torch.compile(backend="aot_eager", fullgraph=True)
def f(x):
    x = torch.ops.aten.detach(x)
    res = x + 1
    return res

f(torch.ones(1))
```

**Comment by bdhirsh** [comment]
How problematic are the duplicates? This makes the graph a bit confusing, but in terms of runtime it matters less since inductor can optimize the duplicate views away.

I vaguely remember from a while ago that the multiple detach calls were an artifact of how we traced through the autograd engine, so figuring out how to avoid tracing them in the first place is probably work

**Comment by StrongerXi** [comment]
> in terms of runtime it matters less since inductor can optimize the duplicate views away.

That's true, but it actually bumps up compilation time quite a bit -- search for `50s in aot-dispatcher` in #150706. And currently the mitigation strategy is either [my Dynamo-level workaround](https://github.com/pytorch/pytorch/pull/151265#discussion_r2042984519), or fixing this.

**Comment by aorenste** [comment]
Classified as a **framework-internal design bug** — the issue describes an architectural problem in PyTorch (dispatcher, codegen, ABI, etc.); any runnable snippet is illustrative of a symptom rather than a reproduction. Auto-triage is skipping this because "no longer reproduces" wouldn't be a valid close signal for an issue like this — the symptom path may have been worked around without fixing the root cause.

Classifier notes: AOTDispatcher emits multiple redundant aten.detach nodes for a single Dynamo detach; an artifact of tracing through the autograd engine. Symptom is graph shape (visible only via TORCH_LOGS aot_graphs), not a runtime error — running the snippet succeeds, so it isn't operationally reproducible without graph introspection.

_(Analysis done by Claude.)_
